File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/entities/DyedVariantBlockEntity.java
package generations.gg.generations.core.generationscore.common.world.level.block.entities;

import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.common.world.level.block.utilityblocks.DyeableBlock;
import generations.gg.generations.core.generationscore.common.world.level.block.utilityblocks.DyeableBlock;
import generations.gg.generations.core.generationscore.common.world.level.block.utilityblocks.DyeableBlock;
import net.minecraft.core.BlockPos;
import net.minecraft.world.item.DyeColor;
import net.minecraft.world.level.block.state.BlockState;
import org.joml.Vector3f;

import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class DyedVariantBlockEntity<T extends DyedVariantBlockEntity<?>> extends ModelProvidingBlockEntity implements ModelContextProviders.TintProvider {
    public DyedVariantBlockEntity(MutableBlockEntityType<T> arg, BlockPos arg2, BlockState arg3) {
        super(arg, arg2, arg3);
    }

    @Override
    public String getVariant() {
        return getColor().getSerializedName();
    }

    public DyeColor getColor() {
        return ((DyeableBlock<?, ?>) getBlockState().getBlock()).getColor();
    }

    @Override
    public Vector3f getTint() {
        return COLOR_MAP.getOrDefault(getColor(), null);
    }

    public static final Map<DyeColor, Vector3f> COLOR_MAP = Stream.of(DyeColor.values()).collect(Collectors.toMap(a -> a, dyeColor -> new Vector3f(dyeColor.getTextureDiffuseColors()[0], dyeColor.getTextureDiffuseColors()[1], dyeColor.getTextureDiffuseColors()[2])));
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/entities/DyedVariantBlockEntity.kt
package generations.gg.generations.core.generationscore.common.world.level.block.entities

import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders.TintProvider
import generations.gg.generations.core.generationscore.common.world.level.block.utilityblocks.DyeableBlock
import net.minecraft.core.BlockPos
import net.minecraft.world.item.DyeColor
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import org.joml.Vector3f

fun Int.toVec3f(): Vector3f = Vector3f(
    ((this shr 16) and 0xFF) / 255f,
    ((this shr 8) and 0xFF) / 255f,
    (this and 0xFF) / 255f
)

abstract class DyedVariantBlockEntity<T : DyedVariantBlockEntity<T>>(
    arg: BlockEntityType<T>,
    arg2: BlockPos,
    arg3: BlockState,
) : ModelProvidingBlockEntity(arg, arg2, arg3), TintProvider {
    override fun getVariant(): String? {
        return (blockState.block as DyeableBlock<*, *>).variant
    }

    val color: DyeColor
        get() = (blockState.block as DyeableBlock<*, *>).color

    override fun getTint(): Vector3f? {
        return COLOR_MAP[color]
    }

    companion object {
        @JvmField
        val COLOR_MAP: Map<DyeColor, Vector3f?> = DyeColor.entries.associateWith { it.textureDiffuseColor.toVec3f() }
    }
}
